### DISCUSSION TITLE
Create a multibuffer from multiple files?

### DISCUSSION THREAD

**Discussion by xpe** [zed-industries/zed#6615] [Feature Requests]
I'd like to choose some files, smoosh them together in a multi-buffer, looking something like this:

```
===== file_1.md =====
# File 1
This is file 1

===== file_2.md =====
# File 2
This is file 2

===== file_3.md =====
# File 3
This is file 3
```

Of course, having a nice visual horizontal separator would be nice.

**Design Questions**

- how to specify the files to include?
    - perhaps files inside a directory
    - perhaps a multi-selection
    - perhaps a newline-separated list of files
- how to specify the order of files?
    - initially
    - later
- adding/removing files to/from the multibuffer
- convenient UX for renaming files in the multi-buffer

**Comment by JosephTLyons** [comment]
I'd also like a command to `open selections in multi-buffer`.  This would be useful for a variety of things, but one would be:

Running `Select All Matches` and then running the `open selections in multi-buffer` command to pull everything into one tight view, where you can easily double check that all of your selections are selecting the things you want

**Reply by xpe** [reply]
Hi, is your "friendly amendment" / suggestion about multiple selections in a single file? Would it work for multiple files?

**Comment by mikayla-maki** [comment]
These features sound awesome!

**Reply by JosephTLyons** [reply]
I've had a branch open for probably 6 months with a bit of work for `Open Selections in Multi-buffer`, but I ran into some issues and abandoned it.  I've been wanting that for a long time now 😂

**Comment by alejandrovrojas** [comment]
@JosephTLyons I've also been hoping https://github.com/zed-industries/community/issues/698 would get some attention at some point! 

@xpe I think a file based selection would quickly lead to a very cluttered multi-buffer. I'd like to use a command like `append selection to multi-buffer` which would solve the issue of what to include and in which order, and most importantly only show the most relevant lines of each file.

On that note, Emacs has a feature called [narrowing](https://www.youtube.com/watch?v=4z-ZAklZOwY) which allows to isolate a buffer to a specific number of lines and thereby limiting all editing operations to those lines only. This feature by itself would be a great addition to Zed but it could lead to an even more interesting usage of a multi-buffer view consisting of narrowed parts of different buffers.

**Reply by e1turin** [reply]
> I'd like to use a command like append selection to multi-buffer which would solve the issue of what to include and in which order, and most importantly only show the most relevant lines of each file.

I have found a way to do similar thing using Bookmarks and `editor: view bookmarks` command. It do not allow to reorder selections but at least I can choose what pieces of code I want to see in multibuffer view. If only it was possible to filter bookmarks by labels...

**Comment by imthatgin** [comment]
This is something I've wanted from vscode for a long time, which has been more or less impossible to implement with just the extension API. Has anyone made progress on this type of feature?

**Reply by SomeoneToIgnore** [reply]
I do no think the plugins should do that, it seems quite fundamental to be included into the core.

The main question here is the UI: how that should be displayed to the user, in particular, how "this buffer and this particular range of it to add" is selected in the interface.

After that is solved, all that's left to do is call 

https://github.com/zed-industries/zed/blob/385e05ee00197a7958fa7c14d94bbf25cb47a1cd/crates/multi_buffer/src/multi_buffer.rs#L1224-L1243

and, optionally, create a new multi buffer before doing that.

**Reply by alejandrovrojas** [reply]
> The main question here is the UI: how that should be displayed to the user, in particular, how "this buffer and this particular range of it to add" is selected in the interface.

Wouldn't a command like `open selections in multi-buffer` be enough to solve this? I don't think UI affordances are necessary to test this concept

**Reply by SomeoneToIgnore** [reply]
Better than nothing, but does not sound very convenient to me personally: specifically, I'd love to see the ability to fill a multi buffer from multiple different files, while `open selections ...` deals with a single file only.

**Reply by SomeoneToIgnore** [reply]
A small update, looks like a team member have been thinking about the same, and now there's https://github.com/zed-industries/zed/pull/23644 coming next week.

So, first step towards multi buffer ergonomics is made.

**Comment by DavidPD** [comment]
Man I've been wanting something similar to this since I was a heavy Sublime Text 2 user. The sublime search buffer always made me feel like I should be able to do exactly what Zed offers with multibuffer editing. Then, VSCode's peek made me think about how it would feel if I could open a buffer for a single function and then open buffers for every function my current buffer _references_.

That functionality is probably deeper than Zed's Core should handle since every language would be a little different (unless there's somehow an LSP call for getting the start and end of a function). But if Zed's plugin API made creating and manipulating multibuffers easy...

Basically I've always wondered what it would be like to have a "focus mode" for individual functions and whatever they depend on. Maybe even the ability to save a named multibuffer with a list of functions used in a particular feature.

**Reply by imthatgin** [reply]
Yeah this is exactly what I want. The video from this neovim plugin that achieves what I want is indicative of the value proposition: https://github.com/hoschi/yode-nvim

**Comment by claytonrcarter** [comment]
Thank you, @JosephTLyons for #23644! I have run into a number of occasions where I have wanted to do normal/single buffer search, but to see all of the matches opened in a multibuffer. I *still* want that, but I now have a pretty good intermediate workaround: start a buffer search, run/click "select all matches", run "open selections in multibuffer". Magic!

If such a hypothetical "open all matches in multibuffer" feature were added to the buffer search, the only difference between what I mentioned about is that it would not "select all matches" first.

Anyway, thank you!

**Reply by JosephTLyons** [reply]
I'd like to introduce a bunch of ways to create your own multibuffers. I added `editor: open selections in multibuffer`, but I'd like to be able to open a multibuffer from selections across multiple panes, I'd like to be able to create a multibuffer by selecting multiple files in the project tree, or right clicking on a directory. Creating one from a buffer search sounds cool too.

**Reply by imthatgin** [reply]
Would it be possible to have a command that pulls all open tabs into a single multi buffer? That'd solve a lot of my daily use cases, I think.

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/community-links

**Comment by xpe** [comment]
Is there some rationale for why Discord is better? I find it overwhelming and distracting. GitHub issues feel _much_ better organized. I would guess that Zed's drive to pare down Github issues has a lot to do with PR. I agree that PR matters, but some of these GitHub issues are substantive enough that I try to check back on them from time to time. I'll never really have a good mechanism to do that with Discord :/
